oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -150,26 +155,118 @@ impl Runtime {
-        for (_, node) in nodes.drain() {
-            node.join_handle.join().expect("Failed to join handle");
+        // it will not be able to enter into a blocking state again.
+
+        let join_handles: Vec<Option<JoinHandle<()>>> = {
+            let mut nodes = self.nodes.write().unwrap();
+            nodes
+                .iter_mut()
+                .map(|(_, node)| node.join_handle.take())
+                .collect()
+        };
+        for join_handle in join_handles {
+            if let Some(h) = join_handle {
+                h.join().expect("Failed to join handle");
+            }
+        }
+    }
+
+    fn track_handles_in_node<'a, I: Iterator<Item = &'a Handle>>(
+        &self,
+        node_id: NodeId,
+        handles: I,
+    ) {
+        if node_id == RUNTIME_NODE_ID {
+            return;
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        if let Some(node) = nodes.get(&node_id) {
+            let mut tracked_handles = node.handles.lock().unwrap();
+            for handle in handles {
+                tracked_handles.insert(*handle);
+            }
+        }
+    }
+
+    fn filter_handle(&self, node_id: NodeId, handle: Handle) -> Option<Handle> {
+        if node_id == RUNTIME_NODE_ID {
+            return Some(handle);
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        if let Some(node) = nodes.get(&node_id) {
+            let tracked_handles = node.handles.lock().unwrap();
+            if tracked_handles.contains(&handle) {
+                Some(handle)
+            } else {
+                error!(
+                    "filter_handle: handle {:?} not found in node {:?}",
+                    handle, node_id
+                );
+                None
+            }
+        } else {
+            error!("filter_handle: node {:?} not found", node_id);
+            None
+        }
+    }
+
+    fn filter_optional_handles<'a, I: Iterator<Item = &'a Option<Handle>>>(
+        &self,
+        node_id: NodeId,
+        handles: I,
+    ) -> Vec<Option<Handle>> {
+        if node_id == RUNTIME_NODE_ID {
+            return handles.copied().collect();
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        if let Some(node) = nodes.get(&node_id) {
+            let tracked_handles = node.handles.lock().unwrap();
+
+            handles
+                .map(|optional_handle| {
+                    optional_handle.and_then(|handle| {
+                        if tracked_handles.contains(&handle) {
+                            Some(handle)
+                        } else {
+                            error!(
+                                "filter_optional_handles: handle {:?} not found in node {:?}",
+                                handle, node_id
+                            );
+                            None
+                        }
+                    })
+                })
+                .collect()
```

Comment:
I have not read carefully this yet, but at first glance it's not obvious to me what's going on here. Hopefully more comments throughout would help, but also it would be good to have some tests.

Reply:
Added more comments. I can also revisit this in the future to be clearer and try to simplify.